packages/chevrotain/src/parse/grammar/interpreter.ts

```diff
@@ -383,13 +379,13 @@ export function nextPossibleTokensAfter(
   })
 
   while (!isEmpty(possiblePaths)) {
-    const currPath = possiblePaths.pop()
+    const currPath = possiblePaths.pop()!
 
     // skip alternatives if no more results can be found (assuming deterministic grammar with fixed lookahead)
     if (currPath === EXIT_ALTERNATIVE) {
       if (
         foundCompletePath &&
-        last(possiblePaths).idx <= minimalAlternativesIndex
+        last(possiblePaths)!.idx <= minimalAlternativesIndex
```

Comment:
if we have to use `!` multiple times on `possiblePaths` maybe it is best to assert it is none-empty.

- is it possible to assert a type guard? or can a type guard only be added via a function invocation?

Reply:
possiblePaths could be empty here as we've just done a .pop() above. That it isn't when currPath === EXIT_ALTERNATIVE is a runtime invariant which we are explicitly asserting with `!` here. It seems the more correct way.

>     * is it possible to assert a type guard? or can a type guard only be added via a function invocation?

I'm not sure what you mean?

Reply:
I now see these non null assertions on pop/last/first as idiomatic